File: src/bot.py
import discord
from discord.ext import commands
from config import DISCORD_TOKEN
from client import ask_client, image_opinion_openai
import asyncio
from personas import OPEN_AI_PERSONA as PERSONA
from constants import DEFAULT_OPENAI_MODEL

intents = discord.Intents.default()
intents.message_content = True

bot = commands.Bot(command_prefix='!', intents=intents, help_command=None)

def format_channel_history(ctx, num_messages):
    messages = []
    async def fetch():
        async for message in ctx.channel.history(limit=num_messages):
            messages.append(f"{message.author.display_name}: {message.content}")
        messages.reverse()
        return "\n".join(messages)
    return fetch

@bot.event
async def on_ready():
    print(f'Logged in as {bot.user} (ID: {bot.user.id})')
    print('------')

@bot.command()
async def ping(ctx):
    """Check if the bot is responsive."""
    await ctx.send('Pong!')

def extract_provider_and_args(args, default_provider="openai"):
    """Helper to extract provider and remaining args from a list of arguments."""
    provider = default_provider
    if args and args[0].lower() in ["grok", "openai"]:
        provider = args[0].lower()
        args = args[1:]
    return provider, args

async def send_long_response(ctx, response):
    """Helper to send long responses in 2000-char chunks."""
    for i in range(0, len(response), 2000):
        await ctx.send(response[i:i+2000])

@bot.command()
async def ask(ctx, *, prompt: str):
    """Ask a question to the bot persona. Usage: !ask [grok|openai] question"""
    await ctx.send('Thinking...')
    args = prompt.split()
    provider, args = extract_provider_and_args(args)
    prompt_text = " ".join(args) if args else prompt
    response = await asyncio.to_thread(
        ask_client, prompt_text, system_message=PERSONA, model=DEFAULT_OPENAI_MODEL, provider=provider
    )
    await send_long_response(ctx, response)

@bot.command()
async def opinion(ctx, *, arg: str = ""):
    """Summarize or form an opinion on recent messages. Usage: !opinion [grok|openai] [num_messages]"""
    await ctx.send('Let me think about what everyone has been saying...')
    args = arg.split()
    provider, args = extract_provider_and_args(args)
    num_messages = 10
    if args and args[0].isdigit():
        num_messages = int(args[0])
    context_str = await format_channel_history(ctx, num_messages)()
    system_message = (
        f"{PERSONA}\nHere are the last {num_messages} messages in this channel:\n{context_str}\n"
        "Form an opinion or summary about the conversation."
    )
    response = await asyncio.to_thread(
        ask_client, "What is your opinion on the recent conversation?", system_message=system_message,
        model=DEFAULT_OPENAI_MODEL, provider=provider
    )
    await send_long_response(ctx, response)

@bot.command()
async def who_won(ctx, *, arg: str = ""):
    """Analyze arguments and determine who won. Usage: !who_won [grok|openai] [num_messages]"""
    await ctx.send('Analyzing the last arguments...')
    args = arg.split()
    provider, args = extract_provider_and_args(args)
    num_messages = 100
    if args and args[0].isdigit():
        num_messages = int(args[0])
    context_str = await format_channel_history(ctx, num_messages)()
    system_message = (
        f"{PERSONA}\nHere are the last {num_messages} messages in this channel:\n{context_str}\n"
        "Based on the arguments and discussions, determine who won the arguments and why. "
        "Be specific and fair, and explain your reasoning."
    )
    response = await asyncio.to_thread(
        ask_client, "Who won the arguments in the recent conversation?", system_message=system_message,
        model=DEFAULT_OPENAI_MODEL, provider=provider
    )
    await send_long_response(ctx, response)

@bot.command()
async def user_opinion(ctx, member: discord.Member, *, arg: str = ""):
    """Form an opinion on a user. Usage: !user_opinion @user [grok|openai] [days] [max_messages]"""
    await ctx.send(f'Analyzing {member.display_name}...')
    from datetime import datetime, timedelta, timezone
    args = arg.split()
    provider, args = extract_provider_and_args(args)
    days = 3
    max_messages = 200
    if args and args[0].isdigit():
        days = int(args[0])
        args = args[1:]
    if args and args[0].isdigit():
        max_messages = int(args[0])
    now = datetime.now(timezone.utc)
    after = now - timedelta(days=days)
    messages = []
    async for message in ctx.channel.history(limit=max_messages, after=after):
        if message.author.id == member.id:
            messages.append(f"{message.author.display_name}: {message.content}")
    if not messages:
        await ctx.send(f"No messages found for {member.display_name} in the last {days} days.")
        return
    context_str = "\n".join(messages)
    system_message = (
        f"Here are all the messages sent by {member.display_name} in the last {days} days in this channel:\n{context_str}\n"
    )
    response = await asyncio.to_thread(
        ask_client, f"What is your opinion of {member.display_name}?", system_message=system_message,
        model=DEFAULT_OPENAI_MODEL, provider=provider
    )
    await send_long_response(ctx, response)

@bot.command()
async def most(ctx, *, question: str):
    """Ask who is the most X or most likely to do Y. Usage: !most [grok|openai] question"""
    num_messages = 100
    await ctx.send(f"Analyzing: {question} (last {num_messages} messages)...")
    args = question.split()
    provider, args = extract_provider_and_args(args)
    question_text = " ".join(args) if args else question
    messages = []
    user_message_count = {}
    async for message in ctx.channel.history(limit=num_messages):
        if message.author.bot:
            continue
        messages.append(f"{message.author.display_name}: {message.content}")
        user_message_count[message.author.display_name] = user_message_count.get(message.author.display_name, 0) + 1
    active_users = sorted(user_message_count.items(), key=lambda x: x[1], reverse=True)[:5]
    active_user_names = [user for user, _ in active_users]
    context_str = "\n".join(messages)
    if len(question_text.split()) == 1:
        prompt = f"Who is the most {question_text} in the recent conversation?"
    else:
        prompt = question_text
    system_message = (
        f"{PERSONA}\nHere are the last {num_messages} messages in this channel:\n{context_str}\n"
        f"Among the most active users ({', '.join(active_user_names)}), answer the following question: {question_text}. "
        f"Explain your reasoning as Coonbot."
    )
    response = await asyncio.to_thread(
        ask_client, prompt, system_message=system_message, model=DEFAULT_OPENAI_MODEL, provider=provider
    )
    await send_long_response(ctx, response)

@bot.command()
async def image_opinion(ctx, *, arg: str = ""):
    """Form an opinion on an image. Usage: !image_opinion [grok|openai] [image_url|prompt]"""
    image_url = None
    custom_prompt = None
    args = arg.split()
    provider, args = extract_provider_and_args(args)
    content_args = ctx.message.content.split()
    # Check for attachment first
    if ctx.message.attachments:
        image_url = ctx.message.attachments[0].url
        if args:
            custom_prompt = " ".join(args)
    # If replying to a message with an image
    elif ctx.message.reference:
        try:
            replied_msg = await ctx.channel.fetch_message(ctx.message.reference.message_id)
            if replied_msg.attachments:
                image_url = replied_msg.attachments[0].url
            if args:
                custom_prompt = " ".join(args)
        except Exception as e:
            await ctx.send(f"Could not fetch replied message: {e}")
            return
    # If no attachment or reply, check for image URL in args
    elif args:
        possible_url = args[0]
        if possible_url.startswith("http://") or possible_url.startswith("https://"):
            image_url = possible_url
            if len(args) > 1:
                custom_prompt = " ".join(args[1:])
        else:
            custom_prompt = " ".join(args)
    if not image_url:
        await ctx.send("Please attach an image, provide a valid image URL (starting with http/https), or reply to a message with an image.")
        return
    await ctx.send("Analyzing image, one sec...")
    try:
        if provider == "grok":
            from client import image_opinion_grok
            opinion = image_opinion_grok(image_url, system_message=PERSONA, custom_prompt=custom_prompt)
        else:
            opinion = image_opinion_openai(image_url, system_message=PERSONA, custom_prompt=custom_prompt)
        await send_long_response(ctx, opinion)
    except Exception as e:
        await ctx.send(f"Error analyzing image: {e}")

@bot.command()
@commands.is_owner()
async def clear_commands(ctx):
    """Clear all global slash commands from Discord."""
    bot.tree.clear_commands(guild=None)
    await bot.tree.sync()
    await ctx.send("Cleared all global slash commands.")

@bot.command()
async def roast(ctx, member: discord.Member = None):
    """Roast a user. Usage: !roast @user or !roast as a reply to a message."""
    target_name = None
    roast_message = None
    # If user is mentioned
    if member:
        target_name = member.display_name
        system_message = f"{PERSONA}\nRoast {target_name} as if you were a boston comedian who grew up in the bronx. Dont be afraid to attack sexuality, its a joke and everyone gets it. Be really really mean."
        prompt = f"Roast {target_name}."
    # If command is a reply to a message
    elif ctx.message.reference:
        try:
            replied_msg = await ctx.channel.fetch_message(ctx.message.reference.message_id)
            target_name = replied_msg.author.display_name
            roast_message = replied_msg.content
            system_message = f"{PERSONA}\nRoast {target_name} based on this message: '{roast_message}'. Be a boston comedian from the bronx, don't be afraid to attack sexuality, it's a joke and everyone gets it."
            prompt = f"Roast {target_name} for saying: {roast_message}"
        except Exception as e:
            await ctx.send(f"Could not fetch replied message: {e}")
            return
    else:
        await ctx.send("Please mention a user or reply to a message to roast.")
        return
    await ctx.send(f"Cooking up a roast for {target_name}...")
    response = await asyncio.to_thread(
        ask_client, prompt, system_message=system_message, model=DEFAULT_OPENAI_MODEL, provider="openai"
    )
    await send_long_response(ctx, response)

def main():
    bot.run(DISCORD_TOKEN)

if __name__ == "__main__":
    main()

File: go/bot.go
package main

import (
	"fmt"
	"log"
	"strconv"
	"strings"
	"time"

	"github.com/bwmarrin/discordgo"
)

// Bot represents the Discord bot
type Bot struct {
	session  *discordgo.Session
	aiClient *AIClient
	config   *Config
}

// NewBot creates a new bot instance
func NewBot(config *Config) (*Bot, error) {
	session, err := discordgo.New("Bot " + config.DiscordToken)
	if err != nil {
		return nil, fmt.Errorf("error creating Discord session: %w", err)
	}

	session.Identify.Intents = discordgo.IntentsGuildMessages | discordgo.IntentsMessageContent

	aiClient := NewAIClient(config.OpenAIAPIKey, config.XAIAPIKey)

	bot := &Bot{
		session:  session,
		aiClient: aiClient,
		config:   config,
	}

	// Register message handler
	session.AddHandler(bot.messageHandler)

	return bot, nil
}

// Start starts the bot
func (b *Bot) Start() error {
	err := b.session.Open()
	if err != nil {
		return fmt.Errorf("error opening connection: %w", err)
	}

	user, err := b.session.User("@me")
	if err != nil {
		return fmt.Errorf("error obtaining account details: %w", err)
	}

	log.Printf("Logged in as %s (ID: %s)", user.Username, user.ID)
	log.Println("------")
	return nil
}

// Close closes the bot session
func (b *Bot) Close() {
	b.session.Close()
}

// messageHandler handles incoming messages
func (b *Bot) messageHandler(s *discordgo.Session, m *discordgo.MessageCreate) {
	// Ignore messages from the bot itself
	if m.Author.ID == s.State.User.ID {
		return
	}

	// Check if message starts with command prefix
	if !strings.HasPrefix(m.Content, "!") {
		return
	}

	// Parse command and arguments
	parts := strings.Fields(m.Content)
	if len(parts) == 0 {
		return
	}

	command := strings.TrimPrefix(parts[0], "!")
	args := parts[1:]

	// Route to appropriate command handler
	switch command {
	case "ping":
		b.handlePing(s, m)
	case "ask":
		b.handleAsk(s, m, args)
	case "opinion":
		b.handleOpinion(s, m, args)
	case "who_won":
		b.handleWhoWon(s, m, args)
	case "user_opinion":
		b.handleUserOpinion(s, m, args)
	case "most":
		b.handleMost(s, m, args)
	case "image_opinion":
		b.handleImageOpinion(s, m, args)
	case "roast":
		b.handleRoast(s, m, args)
	}
}

// handlePing responds with "Pong!"
func (b *Bot) handlePing(s *discordgo.Session, m *discordgo.MessageCreate) {
	s.ChannelMessageSend(m.ChannelID, "Pong!")
}

// extractProviderAndArgs extracts provider from arguments
func extractProviderAndArgs(args []string, defaultProvider string) (string, []string) {
	provider := defaultProvider
	if len(args) > 0 && (strings.ToLower(args[0]) == "grok" || strings.ToLower(args[0]) == "openai") {
		provider = strings.ToLower(args[0])
		args = args[1:]
	}
	return provider, args
}

// sendLongResponse sends long responses in 2000-char chunks
func (b *Bot) sendLongResponse(s *discordgo.Session, channelID, response string) {
	for i := 0; i < len(response); i += 2000 {
		end := i + 2000
		if end > len(response) {
			end = len(response)
		}
		s.ChannelMessageSend(channelID, response[i:end])
	}
}

// handleAsk handles the !ask command
func (b *Bot) handleAsk(s *discordgo.Session, m *discordgo.MessageCreate, args []string) {
	if len(args) == 0 {
		s.ChannelMessageSend(m.ChannelID, "Usage: !ask [grok|openai] <question>")
		return
	}

	s.ChannelMessageSend(m.ChannelID, "Thinking...")

	provider, args := extractProviderAndArgs(args, "openai")
	prompt := strings.Join(args, " ")

	response, err := b.aiClient.AskClient(prompt, OpenAIPersona, DefaultOpenAIModel, provider, DefaultMaxTokens)
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error: %v", err))
		return
	}

	b.sendLongResponse(s, m.ChannelID, response)
}

// formatChannelHistory fetches and formats recent messages
func (b *Bot) formatChannelHistory(s *discordgo.Session, channelID string, numMessages int) (string, error) {
	messages, err := s.ChannelMessages(channelID, numMessages, "", "", "")
	if err != nil {
		return "", err
	}

	// Reverse the messages to show oldest first
	var formatted []string
	for i := len(messages) - 1; i >= 0; i-- {
		msg := messages[i]
		member, err := s.GuildMember(msg.GuildID, msg.Author.ID)
		displayName := msg.Author.Username
		if err == nil && member.Nick != "" {
			displayName = member.Nick
		}
		formatted = append(formatted, fmt.Sprintf("%s: %s", displayName, msg.Content))
	}

	return strings.Join(formatted, "\n"), nil
}

// handleOpinion handles the !opinion command
func (b *Bot) handleOpinion(s *discordgo.Session, m *discordgo.MessageCreate, args []string) {
	s.ChannelMessageSend(m.ChannelID, "Let me think about what everyone has been saying...")

	provider, args := extractProviderAndArgs(args, "openai")
	numMessages := 10

	if len(args) > 0 {
		if n, err := strconv.Atoi(args[0]); err == nil {
			numMessages = n
		}
	}

	contextStr, err := b.formatChannelHistory(s, m.ChannelID, numMessages)
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error fetching messages: %v", err))
		return
	}

	systemMessage := fmt.Sprintf("%s\nHere are the last %d messages in this channel:\n%s\n"+
		"Form an opinion or summary about the conversation.", OpenAIPersona, numMessages, contextStr)

	response, err := b.aiClient.AskClient("What is your opinion on the recent conversation?",
		systemMessage, DefaultOpenAIModel, provider, DefaultMaxTokens)
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error: %v", err))
		return
	}

	b.sendLongResponse(s, m.ChannelID, response)
}

// handleWhoWon handles the !who_won command
func (b *Bot) handleWhoWon(s *discordgo.Session, m *discordgo.MessageCreate, args []string) {
	s.ChannelMessageSend(m.ChannelID, "Analyzing the last arguments...")

	provider, args := extractProviderAndArgs(args, "openai")
	numMessages := 100

	if len(args) > 0 {
		if n, err := strconv.Atoi(args[0]); err == nil {
			numMessages = n
		}
	}

	contextStr, err := b.formatChannelHistory(s, m.ChannelID, numMessages)
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error fetching messages: %v", err))
		return
	}

	systemMessage := fmt.Sprintf("%s\nHere are the last %d messages in this channel:\n%s\n"+
		"Based on the arguments and discussions, determine who won the arguments and why. "+
		"Be specific and fair, and explain your reasoning.", OpenAIPersona, numMessages, contextStr)

	response, err := b.aiClient.AskClient("Who won the arguments in the recent conversation?",
		systemMessage, DefaultOpenAIModel, provider, DefaultMaxTokens)
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error: %v", err))
		return
	}

	b.sendLongResponse(s, m.ChannelID, response)
}

// handleUserOpinion handles the !user_opinion command
func (b *Bot) handleUserOpinion(s *discordgo.Session, m *discordgo.MessageCreate, args []string) {
	if len(args) == 0 {
		s.ChannelMessageSend(m.ChannelID, "Usage: !user_opinion @user [grok|openai] [days] [max_messages]")
		return
	}

	// Parse mentioned user
	if len(m.Mentions) == 0 {
		s.ChannelMessageSend(m.ChannelID, "Please mention a user to analyze.")
		return
	}
	targetUser := m.Mentions[0]

	s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Analyzing %s...", targetUser.Username))

	// Remove the mention from args
	argsWithoutMention := []string{}
	for _, arg := range args {
		if !strings.HasPrefix(arg, "<@") {
			argsWithoutMention = append(argsWithoutMention, arg)
		}
	}

	provider, remainingArgs := extractProviderAndArgs(argsWithoutMention, "openai")
	days := 3
	maxMessages := 200

	if len(remainingArgs) > 0 {
		if n, err := strconv.Atoi(remainingArgs[0]); err == nil {
			days = n
			remainingArgs = remainingArgs[1:]
		}
	}

	if len(remainingArgs) > 0 {
		if n, err := strconv.Atoi(remainingArgs[0]); err == nil {
			maxMessages = n
		}
	}

	// Fetch messages from the user
	after := time.Now().Add(-time.Duration(days) * 24 * time.Hour)
	allMessages, err := s.ChannelMessages(m.ChannelID, maxMessages, "", "", "")
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error fetching messages: %v", err))
		return
	}

	var userMessages []string
	for _, msg := range allMessages {
		if msg.Author.ID == targetUser.ID && msg.Timestamp.After(after) {
			member, err := s.GuildMember(msg.GuildID, msg.Author.ID)
			displayName := msg.Author.Username
			if err == nil && member.Nick != "" {
				displayName = member.Nick
			}
			userMessages = append(userMessages, fmt.Sprintf("%s: %s", displayName, msg.Content))
		}
	}

	if len(userMessages) == 0 {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("No messages found for %s in the last %d days.", targetUser.Username, days))
		return
	}

	contextStr := strings.Join(userMessages, "\n")
	systemMessage := fmt.Sprintf("Here are all the messages sent by %s in the last %d days in this channel:\n%s\n",
		targetUser.Username, days, contextStr)

	response, err := b.aiClient.AskClient(fmt.Sprintf("What is your opinion of %s?", targetUser.Username),
		systemMessage, DefaultOpenAIModel, provider, DefaultMaxTokens)
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error: %v", err))
		return
	}

	b.sendLongResponse(s, m.ChannelID, response)
}

// handleMost handles the !most command
func (b *Bot) handleMost(s *discordgo.Session, m *discordgo.MessageCreate, args []string) {
	if len(args) == 0 {
		s.ChannelMessageSend(m.ChannelID, "Usage: !most [grok|openai] <question>")
		return
	}

	numMessages := 100
	s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Analyzing: %s (last %d messages)...", strings.Join(args, " "), numMessages))

	provider, args := extractProviderAndArgs(args, "openai")
	question := strings.Join(args, " ")

	// Fetch messages and count by user
	allMessages, err := s.ChannelMessages(m.ChannelID, numMessages, "", "", "")
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error fetching messages: %v", err))
		return
	}

	var messages []string
	userMessageCount := make(map[string]int)

	for i := len(allMessages) - 1; i >= 0; i-- {
		msg := allMessages[i]
		if msg.Author.Bot {
			continue
		}

		member, err := s.GuildMember(msg.GuildID, msg.Author.ID)
		displayName := msg.Author.Username
		if err == nil && member.Nick != "" {
			displayName = member.Nick
		}

		messages = append(messages, fmt.Sprintf("%s: %s", displayName, msg.Content))
		userMessageCount[displayName]++
	}

	// Get top 5 active users
	type userCount struct {
		name  string
		count int
	}
	var counts []userCount
	for name, count := range userMessageCount {
		counts = append(counts, userCount{name, count})
	}

	// Simple sort by count
	for i := 0; i < len(counts); i++ {
		for j := i + 1; j < len(counts); j++ {
			if counts[j].count > counts[i].count {
				counts[i], counts[j] = counts[j], counts[i]
			}
		}
	}

	activeUserNames := []string{}
	for i := 0; i < 5 && i < len(counts); i++ {
		activeUserNames = append(activeUserNames, counts[i].name)
	}

	contextStr := strings.Join(messages, "\n")

	prompt := question
	if len(strings.Fields(question)) == 1 {
		prompt = fmt.Sprintf("Who is the most %s in the recent conversation?", question)
	}

	systemMessage := fmt.Sprintf("%s\nHere are the last %d messages in this channel:\n%s\n"+
		"Among the most active users (%s), answer the following question: %s. "+
		"Explain your reasoning as Coonbot.", OpenAIPersona, numMessages, contextStr,
		strings.Join(activeUserNames, ", "), question)

	response, err := b.aiClient.AskClient(prompt, systemMessage, DefaultOpenAIModel, provider, DefaultMaxTokens)
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error: %v", err))
		return
	}

	b.sendLongResponse(s, m.ChannelID, response)
}

// handleImageOpinion handles the !image_opinion command
func (b *Bot) handleImageOpinion(s *discordgo.Session, m *discordgo.MessageCreate, args []string) {
	var imageURL string
	var customPrompt *string

	provider, args := extractProviderAndArgs(args, "openai")

	// Check for attachment first
	if len(m.Attachments) > 0 {
		imageURL = m.Attachments[0].URL
		if len(args) > 0 {
			prompt := strings.Join(args, " ")
			customPrompt = &prompt
		}
	} else if m.MessageReference != nil {
		// If replying to a message
		refMsg, err := s.ChannelMessage(m.ChannelID, m.MessageReference.MessageID)
		if err != nil {
			s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Could not fetch replied message: %v", err))
			return
		}
		if len(refMsg.Attachments) > 0 {
			imageURL = refMsg.Attachments[0].URL
		}
		if len(args) > 0 {
			prompt := strings.Join(args, " ")
			customPrompt = &prompt
		}
	} else if len(args) > 0 {
		// Check for image URL in args
		possibleURL := args[0]
		if strings.HasPrefix(possibleURL, "http://") || strings.HasPrefix(possibleURL, "https://") {
			imageURL = possibleURL
			if len(args) > 1 {
				prompt := strings.Join(args[1:], " ")
				customPrompt = &prompt
			}
		} else {
			prompt := strings.Join(args, " ")
			customPrompt = &prompt
		}
	}

	if imageURL == "" {
		s.ChannelMessageSend(m.ChannelID, "Please attach an image, provide a valid image URL (starting with http/https), or reply to a message with an image.")
		return
	}

	s.ChannelMessageSend(m.ChannelID, "Analyzing image, one sec...")

	var opinion string
	var err error

	if provider == "grok" {
		opinion, err = b.aiClient.ImageOpinionGrok(imageURL, OpenAIPersona, customPrompt)
	} else {
		opinion, err = b.aiClient.ImageOpinionOpenAI(imageURL, OpenAIPersona, DefaultOpenAIVisionModel, DefaultMaxTokens, customPrompt)
	}

	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error analyzing image: %v", err))
		return
	}

	b.sendLongResponse(s, m.ChannelID, opinion)
}

// handleRoast handles the !roast command
func (b *Bot) handleRoast(s *discordgo.Session, m *discordgo.MessageCreate, args []string) {
	var targetName string
	var roastMessage string
	var systemMessage string
	var prompt string

	// If user is mentioned
	if len(m.Mentions) > 0 {
		targetUser := m.Mentions[0]
		member, err := s.GuildMember(m.GuildID, targetUser.ID)
		targetName = targetUser.Username
		if err == nil && member.Nick != "" {
			targetName = member.Nick
		}

		systemMessage = fmt.Sprintf("%s\nRoast %s as if you were a boston comedian who grew up in the bronx. "+
			"Dont be afraid to attack sexuality, its a joke and everyone gets it. Be really really mean.", OpenAIPersona, targetName)
		prompt = fmt.Sprintf("Roast %s.", targetName)
	} else if m.MessageReference != nil {
		// If command is a reply to a message
		refMsg, err := s.ChannelMessage(m.ChannelID, m.MessageReference.MessageID)
		if err != nil {
			s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Could not fetch replied message: %v", err))
			return
		}

		member, err := s.GuildMember(m.GuildID, refMsg.Author.ID)
		targetName = refMsg.Author.Username
		if err == nil && member.Nick != "" {
			targetName = member.Nick
		}
		roastMessage = refMsg.Content

		systemMessage = fmt.Sprintf("%s\nRoast %s based on this message: '%s'. "+
			"Be a boston comedian from the bronx, don't be afraid to attack sexuality, it's a joke and everyone gets it.",
			OpenAIPersona, targetName, roastMessage)
		prompt = fmt.Sprintf("Roast %s for saying: %s", targetName, roastMessage)
	} else {
		s.ChannelMessageSend(m.ChannelID, "Please mention a user or reply to a message to roast.")
		return
	}

	s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Cooking up a roast for %s...", targetName))

	response, err := b.aiClient.AskClient(prompt, systemMessage, DefaultOpenAIModel, "openai", DefaultMaxTokens)
	if err != nil {
		s.ChannelMessageSend(m.ChannelID, fmt.Sprintf("Error: %v", err))
		return
	}

	b.sendLongResponse(s, m.ChannelID, response)
}
